src/Features/ExtendedTranslucency.cpp

```diff
@@ -0,0 +1,160 @@
+#include "ExtendedTranslucency.h"
+
+#include "../Util.h"
+
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(
+	ExtendedTranslucency::MaterialParams,
+	AlphaMode,
+	AlphaReduction,
+	AlphaSoftness);
+
+const RE::BSFixedString ExtendedTranslucency::NiExtraDataName = "AnisotropicAlphaMaterial";
+
+ExtendedTranslucency* ExtendedTranslucency::GetSingleton()
+{
+	static ExtendedTranslucency singleton;
+	return &singleton;
+}
+
+void ExtendedTranslucency::SetupResources()
+{
+	// Default material model buffer only changes in settings UI
+	materialDisableCB.emplace(ConstantBufferDesc<MaterialParams>());
+	MaterialParams disabled{ 0, 0.f, 0.f, 0 };
+	materialDisableCB->Update(disabled);
+	// Default material model buffer only changes in settings UI
+	materialDefaultCB.emplace(ConstantBufferDesc<MaterialParams>());
+	materialDefaultCB->Update(settings);
+	// Dynamic material model buffer changes when we read materials
+	materialDynamicCB.emplace(ConstantBufferDesc<MaterialParams>(true));
+}
+
+void ExtendedTranslucency::BSLightingShader_SetupGeometry(RE::BSRenderPass* pass)
+{
+	auto renderer = RE::BSGraphics::Renderer::GetSingleton();
+	auto* transcluency = ExtendedTranslucency::GetSingleton();
+
+	ID3D11DeviceContext* context = reinterpret_cast<ID3D11DeviceContext*>(renderer->GetRuntimeData().context);
+	ID3D11Buffer* buffers[1];
+	if (auto* data = pass->geometry->GetExtraData(NiExtraDataName)) {
+		auto params = transcluency->settings;
+		const RE::NiRTTI* integerDataRTTI = REL::Relocation<const RE::NiRTTI*>{ RE::NiIntegerExtraData::Ni_RTTI }.get();
+		if (data->GetRTTI() == integerDataRTTI) {
+			params.AlphaMode = static_cast<RE::NiIntegerExtraData*>(data)->value;
+		} else {
+			params.AlphaMode = std::to_underlying(ExtendedTranslucency::MaterialModel::Default);
+		}
+		params.AlphaReduction = 0.f;
+
+		buffers[0] = materialDynamicCB->CB();
+		materialDynamicCB->Update(&params, sizeof(params));
+		context->PSSetConstantBuffers(materialCBIndex, 1, buffers);
```

Comment:
You could use hook into Unmap in order to inject parameters into vanilla PerGeometry buffer
https://github.com/doodlum/skyrim-community-shaders/blob/dev/src/Hooks.cpp#L187

Reply:
I end up with a truly constant constant buffers solution:
We have one cbuffer for each material model, and it never needs runtime update: We just assign them into the shader slot. This should avoid most of the cost of synchronized GPU access ;)

BTW, based on the name of these cbuffer, I thought Skyrim is having one DX constant buffer per-geometry/material based on CRC, or we only have one cbuffer for each shader, and Map it every frame?